feat: Intelligent dynamic auto-complete for entering tasks

esm7 opened the pull request:
## Description

This PR adds a dynamic, intelligent `EditorSuggest` menu that makes it super easy & fast to enter tasks inside notes.
If a task structure is detected in the current line, and a suggest request is initiated by Obsidian, it tries to provide the most relevant suggestions like...

- Task components (the various emojis).
- Options for a date in the case the cursor is in a due date, start date or schedule date section.
- Options for a recurrence string in case the cursor is in a recurrence section.

This is activated only when a corresponding setting is on.

Also...

- To do this without increasing the code duplication, I grouped all the various emojis into one place in `Task.ts` and used it from there across the entire plugin.
- I added support for combining task tags with other task components, e.g. `task description #t/city 🔽 #t/next_action`, because when this feature is used to freely enter task content this comes more naturally than requiring tags to be before all other task components.

## Motivation and Context

I felt that using the dialog to enter tasks is not always the right solution for me, especially when I want to quick-capture ideas, but entering tasks as free text is error prone, and entering the emojis is cumbersome for me even if some external mapping is used.
With this change, tasks can be entered freely and fluently right from within the editor!

## How has this been tested?

I've been using Obsidian Tasks with this code to manually migrate 100+ tasks from my previous system, and posted this PR only when I have gained enough confidence that it's incredibly useful and stable.
Also, I used the existing dialog here and there to verify no compatibility is broken, and of course ran all the tests.

## Screenshots (if appropriate):

![Peek 2022-06-22 20-29](https://user-images.githubusercontent.com/10722656/175102574-78b0f851-cc48-4255-a40e-d3036bec5bb6.gif)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Refactor (non-breaking change which only improves the design or structure of existing code, and making no changes to its external behaviour)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project and passes `yarn run lint`.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] My change has adequate Unit Test coverage.

By creating a Pull Request you agree to our [Code of Conduct](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CODE_OF_CONDUCT.md). For further guidance on contributing please see [contributing guide](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CONTRIBUTING.md)

claremacrae commented:
Wow this looks amazing, from the gif. Thank ou very much indeed.

I'm only on an ipad at the moment so cannot currently test it out. I like the refactoring to remove the repetition of symbols in code.

You checked the box to say that there are enough tests, but I don't see any changes in an test files. It looks like there is new behaviour that has been added to Task, so it would be good to see tests to exercise that behaviour, to avoid the need to manually test the behaviour repeatedly in future.

What devices have you tested it on please? Any mobile devices?

esm7 commented:
Actually when thinking about it, you're right that the small new behavior I added to `Task` is not covered by the tests. I'll add it.
Also, I was originally thinking that the completion itself is too UI-centric to add tests for, but actually I should, and will do that too.
Regarding mobile devices, I'm using it on my Android phone and that's where it really shines. The suggestions make entering tasks on the phone a breeze! Would love to know how it works on your ipad :)

claremacrae commented:
Thanks!

If you have problems automating the tests and would like a hand, I’m happy to work together on it. It’s kind of my specialty….

claremacrae commented:
Ooh, another question. Does it work if the global filter is turned off?

claremacrae commented:
> Ooh, another question. Does it work if the global filter is turned off?

I’d be happy to release it without that, and then add it later…. Just wanted to know. Especially as someone who doesn’t use a global filter!

esm7 commented:
By global filter turned off you mean an empty string as the global filter? It decides if suggestions need to be triggered by --

```
        if (
            line.contains(this.settings.globalFilter) &&
            line.match(task.Task.taskRegex)
        )
```
So an empty string should always match.

esm7 commented:
> Thanks!
> 
> If you have problems automating the tests and would like a hand, I’m happy to work together on it. It’s kind of my specialty….

I think the tests won't pose an issue, but if you have time to start reviewing the code itself before I manage to work on the tests (which will either be today or Sunday), it can be great to save some time :)

claremacrae commented:
> I think the tests won't pose an issue, but if you have time to start reviewing the code itself before I manage to work on the tests (which will either be today or Sunday), it can be great to save some time :)

thanks.

Sorry, I won’t be able to do any level of code review today unfortunately.

esm7 commented:
Updated the PR.

1. Added the missing test to `Task` for covering the extended functionality.
2. Added multiple tests to the suggestion generation mechanism.

claremacrae commented:
Hi @esm7, I've spent 2 or 3 hours experimenting with this, reading the code, running the tests, trying to break the tests. Thank you again!

### Changes in behaviour

I do think the auto-complete functionality would be absolutely fine to release as is - huge thanks for doing this work. I had no idea this was even possible.

The area that I am not (yet) comfortable with is:

> I added support for combining task tags with other task components, e.g. task description #t/city 🔽 #t/next_action, because when this feature is used to freely enter task content this comes more naturally than requiring tags to be before all other task components.

My concerns, for your consideration, are around:

1. I worry that this will cause users to accidentally create tasks that incorrectly are now found by searches. This is because it trains them to get used to allowing non-special characters to be after the dates... One accidental deletion of the `#` in text after the dates, and all the priority and date info is not parsed by Tasks.
2. I don't understand the comment 'this comes more naturally than requiring tags to be before all other task components.' Why?
3. I don't know the rest of the tasks code well enough to understand the possible other consequences of this change
4. There are multiple PRs in flight that allow the global filter to be at the end of the line. It's not clear to me how that will interact with this change.

### Suggestions - not yet

I made a note of a few small changes in behaviour that I think would be helpful over the longer run, but:

1.  I don't think any of them should stop this first version being released in to the hands of users, as it is **so** useful!
4. I would be interested to see what improvements other users suggest

### Code review

I'll make a few code review comments now. Depending on how much time you have available, I would be happy to implement some of them myself later in, if you didn't have much time to spare.

esm7 commented:
Hey,

Thanks for taking the time to thoroughly review this!

Please go ahead with any suggestions, pure code or functionality. I'm planning to spend a few more hours on this plugin next week, so I can definitely do some (though I'm hoping to use the time to write another feature that I'm missing -- `OR` queries).

Regarding allowing tags to come after task sections (is there a correct term better than task sections?).
The thing is, that in the current (before this PR) state of obsidian-tasks, a reasonable user cannot really reach a state of a malformed task line, because they would normally only use the dialog to add or edit tasks.
What I was trying to do in this PR, is to make task-entering as natural as writing anything else in Obsidian. However, under this flow, there is no way for the user to know that some ordering of the components on their task is ok and some is not. They can enter tasks intuitively, and then discover "the hard way" (when they're not found by queries) that something is wrong in the order of the line.

I'm indicating this out of my own experience: after writing this feature, and entering over 100 tasks, I found "the hard way" that some of them aren't shown in my queries... I had to debug quite a bit (originally thinking I broke something) just to discover that all the tasks on which I added tags at the end are actually not parsed by obsidian-tasks :sweat: 

Ideally I think we should allow a completely free order of anything in the task line, but this is very challenging from a parsing perspective and very error-prone (see for example how Todoist parses tasks "naturally", and is often wrong -- but you can immediately see this in the UI because the components the due date are colored differently while typing).
So since a completely free order is not really feasible right now, I think the least we can do is to allow to mix tags with other components. It's easy to parse and Obsidian already highlights tags differently than the text. It's now also properly covered with tests, so hopefully future PRs cannot break anything that relates to this functionality.

If you think this does more damage than good, I see two alternatives:
1. Have a setting of "strict ordering" (we can keep this the default) and "loose ordering".
2. I can add an indication for whether the current line is a malformed task, e.g. in the Obsidian status bar. It can even be taken a step forward and have the 'Enter' key in the suggestion popup reorganize the task line -- but I believe users will find it obstructive.

claremacrae commented on src/Suggestor/Suggestor.ts:
```diff
@@ -0,0 +1,379 @@
+import * as task from '../Task';
+
+const datePrefixCharacters = `${task.startDateSymbol}${task.scheduledDateSymbol}${task.dueDateSymbol}`;
+const maxTotalSuggestions = 6;
```
In my copy, I increased this a lot. I thought it was better to see more suggestions and experiment and find that I could filter the list, rather than not be aware that there were other options I could search for.

esm7 replied:
Do you think this should be configurable?
I'm personally not in favor of too many configuration options, but maybe this max value is too much of a personal taste.
A higher value has its benefits like you mention, but a smaller value (and thus a smaller-sized menu) is less intrusive when typing, and much more feasible for mobile.

claremacrae replied:
I think making it configurable would be good then. Though having 3 or 4 configurations for this one feature and having to restart after making every change, to test it, might get a bit old.

claremacrae commented on src/Task.ts:
```diff
@@ -231,6 +245,8 @@ export class Task {
         let recurrence: Recurrence | null = null;
         let tags: any = [];
+        // Tags that are removed from the end while parsing, but we want to add them back for being part of the description
+        let trailingTags = '';
```
Minor observation:

This is called `trailingTags` but I found that it parsed this and retained all the values...

`- [ ] this is a task due 📅 2021-09-12 #inside_tag ⏫ #some/tags_with_underscore ⏳ 2021-12-28 #more_tags`

So maybe `trailing` is a bit confusing?

esm7 replied:
True, "trailing" here actually means "after any description text" and not "after everything", which may not be the best name.

claremacrae commented on src/config/SettingsTab.ts:
```diff
@@ -78,5 +78,37 @@ export class SettingsTab extends PluginSettingTab {
+
+        new Setting(containerEl)
+            .setName('Auto-suggest task content')
+            .setDesc(
+                'Enabling this will open an intelligent suggest window while typing inside a recognized task line',
+            )
```
I wonder about adding `in Live Preview  and Source mode.` to the message?

It's great that it works in both.

claremacrae reviewed (approved):
I'm approving this, as if you didn't have any more time to work on it, I would definitely go ahead and merge it, and then probably make some of the suggestions myself. :-)

I would still like to understand a bit better the motivation for including the tags changes - And I would generally prefer such a change to be in a separate PR, for reasons including:

- it is really independent of the suggestions feature, and deserves a PR mention in its own right, such as in release notes
- including it here means that more docs need to be written before we can release the suggestion feature
- There is someone I would like to ask to review the tags changes, but right now, they would have to also review all the suggestion code to spot the tags changes...

claremacrae commented:
> Hey,
> 
> Thanks for taking the time to thoroughly review this!
> 
> Please go ahead with any suggestions, pure code or functionality. I'm planning to spend a few more hours on this plugin next week, so I can definitely do some (though I'm hoping to use the time to write another feature that I'm missing -- `OR` queries).

OK, I would rather you spent time on the `OR` queries than furtle with the code improvements I mentioned in the code review. If you could answer the questions I asked, that would be great.

Please do tell us what you have in mind for queries before diving in to it. You've probably seen the suggestions in https://github.com/obsidian-tasks-group/obsidian-tasks/discussions/121 - so that might be a good place to have the discussion?

> Regarding allowing tags to come after task sections (is there a correct term better than task sections?). The thing is, that in the current (before this PR) state of obsidian-tasks, a reasonable user cannot really reach a state of a malformed task line, because they would normally only use the dialog to add or edit tasks. 

Oh that's interesting. I've answered support questions from users who are definitely typing tasks up by hand, but I do not know their proportion of the whole.

> What I was trying to do in this PR, is to make task-entering as natural as writing anything else in Obsidian. However, under this flow, there is no way for the user to know that some ordering of the components on their task is ok and some is not. They can enter tasks intuitively, and then discover "the hard way" (when they're not found by queries) that something is wrong in the order of the line.

I do understand what you are saying... What I'm unsure of is, why just tags?

For example someone could write:

`- [ ] this is a task due #inside_tag #some/tags_with_underscore #more_tags ⏫ ⏳ 2021-12-28 then talk to Georgina 📅 2021-09-12`

And the `then talk to Georgina` would also break the parsing. So I don't understand why tags deserve special treatment here.

> I'm indicating this out of my own experience: after writing this feature, and entering over 100 tasks, I found "the hard way" that some of them aren't shown in my queries... I had to debug quite a bit (originally thinking I broke something) just to discover that all the tasks on which I added tags at the end are actually not parsed by obsidian-tasks 😓

I understand.

> Ideally I think we should allow a completely free order of anything in the task line, but this is very challenging from a parsing perspective and very error-prone (see for example how Todoist parses tasks "naturally", and is often wrong -- but you can immediately see this in the UI because the components the due date are colored differently while typing). So since a completely free order is not really feasible right now, I think the least we can do is to allow to mix tags with other components. It's easy to parse and Obsidian already highlights tags differently than the text. It's now also properly covered with tests, so hopefully future PRs cannot break anything that relates to this functionality.
> 
> If you think this does more damage than good, I see two alternatives:
> 
> 1. Have a setting of "strict ordering" (we can keep this the default) and "loose ordering".
> 2. I can add an indication for whether the current line is a malformed task, e.g. in the Obsidian status bar. It can even be taken a step forward and have the 'Enter' key in the suggestion popup reorganize the task line -- but I believe users will find it obstructive.

OK let's go for it. We can decide later how to show up tasks that look like they have symbols in that should have been interpreted, of which your status option is one idea. (Although it isn't visible on mobile)

However, I would still prefer the tags change to be a separate PR please. Please tell me if you would prefer me to separate it out from the current PR....

esm7 commented on src/Task.ts:
```diff
@@ -87,6 +100,7 @@ export class Task {
     public static readonly dueDateRegex = /[📅📆🗓] ?(\d{4}-\d{2}-\d{2})$/u;
     public static readonly doneDateRegex = /✅ ?(\d{4}-\d{2}-\d{2})$/u;
     public static readonly recurrenceRegex = /🔁 ?([a-zA-Z0-9, !]+)$/iu;
+    public static readonly tagsRegex = /(#[\p{L}\p{N}_/-]*\s*)+$/u;
```
That's a unicode regex, `\p{L}\p{N}_` is equivalent to `\w`. It's a different approach to `hashTags` which defines tags by negation (but is currently not unicode, so I assume tags in many languages don't work).
It's possibly wiser though to use the same regex for tags in both places.

esm7 commented:
1. I'll take out the tags functionality to a separate PR, that's not a problem, if it makes you more comfortable it's totally worth it.
2. Actually, thinking about it more thoroughly, the only actual limitation in "fully free" task ordering is the recurrence rule. So maybe in that PR I can go a step forward and allow all the components of the task line to be mixed except the recurrence rule... What do you think? IMHO it's an important extension of the task syntax for the day after auto-complete is widely used, and users may want to write tasks in free-form with minimal limitations. (I assume the support requests that you mentioned come from more advanced users, who have mapped the emojis to keyboard shortcuts. Imagine *all* users trying to type arbitrary tasks without realizing they won't be parsed by the plugin...)
3. Will comment in https://github.com/obsidian-tasks-group/obsidian-tasks/discussions/121 about what I had in mind (and believe I can implement rather quickly).

claremacrae commented:
> * I'll take out the tags functionality to a separate PR, that's not a problem, if it makes you more comfortable it's totally worth it.

Many thanks. I take your point about wanting to release the two together. Happy to wait until both are ready before releasing....

> * Actually, thinking about it more thoroughly, the only actual limitation in "fully free" task ordering is the recurrence rule. So maybe in that PR I can go a step forward and allow all the components of the task line to be mixed except the recurrence rule... What do you think? IMHO it's an important extension of the task syntax for the day after auto-complete is widely used, and users may want to write tasks in free-form with minimal limitations.

Yes it is recurrence which is the particular challenge.

How might you be able to allow complete free form, and still allow recurrence?  This sounds intriguing.

> (I assume the support requests that you mentioned come from more advanced users, who have mapped the emojis to keyboard shortcuts. Imagine _all_ users trying to type arbitrary tasks without realizing they won't be parsed by the plugin...)

One was about how to make it easier to add the special symbols - which you have nailed.

Another was because their task was not found because they added something after the symbols. (I think it was a tag, and yes, I do appreciate the irony in admitting this!!!!)

> * Will comment in [Filter combination (OR) #121](https://github.com/obsidian-tasks-group/obsidian-tasks/discussions/121) about what I had in mind (and believe I can implement rather quickly).

Super - thank you!

esm7 commented:
Done: this PR now includes solely the auto-complete functionality and its tests (and of course also the little refactoring to the emojis).
I will open a new "loose ordering" PR for the other functionality which I believe should be important given the understanding that with auto-complete, many more users will want to type their task rather than edit them in the dialog, and a strict order requirement that silently fails is problematic under this flow.

esm7 commented:
There are some comments you wrote that I did not reply or treat, I hope it's indeed ok for me to move on for now, I still have two PRs to work on around here :-)

claremacrae commented:
@esm7 I was thinking that I couldn't merge this until your parsing improvements landed, to prevent the presence of this blocking me doing a release....

But then I wondered, could I merge it, and then comment out the changes in main that activate it.

This would remove the chances of future changes conflicting with this one, and allow other changes too take advantage of your refactoring...

So I think that is what I'll go ahead and do, soon...

esm7 commented:
My parsing changes are actually complete and only waiting for this PR to be merged :) I based them on this branch because it was easier than testing it completely independently (even though in theory there shouldn't be a relation). So I'm not opening a PR yet because I don't want the diff to show as the two PRs together.

claremacrae commented:
I've tested it on Mac, iPhone and iPad and it works great on them all.

claremacrae commented:
> My parsing changes are actually complete and only waiting for this PR to be merged :) I based them on this branch because it was easier than testing it completely independently (even though in theory there shouldn't be a relation). So I'm not opening a PR yet because I don't want the diff to show as the two PRs together.

Well that was a question worth asking! No problem...

I'll leave the conversations unresolved for now, to come back to after the other PR has merged, to avoid risking any conflicts with your other changes.

esm7 commented:
@AnnaKornfeldSimpson it refers to anything that is matched from the auto-complete menu, basically anything that it is able to show.
e.g. the word 'due', 'repeat', 'recurr', 'priority', 'low', 'high' etc, and of course more specific completions -- e.g. various dates and phrases that you can see in the menu when you enter the 'due' or 'recurr' emoji and start typing. A good start would be to use it with length 0 for a while, get used to a few flows (e.g. I'm now very used to type 'task description... du<auto-completes> tom<auto-completes>'), and then when you get familiar with what's possible increase the match length to something higher so it would be less obstructive.
@claremacrae let me know if you think this is something that should be elaborated in the settings description, and if so how.

esm7 commented:
Indeed, I started using it with a min length of 2, so it very rarely flashes unexpectedly when entering a task. (Either way, it will not pop up outside of a task line that includes the global filter.)
The idea here is that you can fill a whole task without ever leaving the line you are typing :)

claremacrae commented:
This was a helpful discussion to see.

I feel that this needs a page of docs in its own right to explain, from a user perspective, how it works, how to take advantage of it and what the settings do.

Without that, I worry about the level of user repeated support questions it could generate.

I've already got a list of small things that I think people will ask about, and I want to nip them in the bud...

claremacrae commented:
> I've already got a list of small things that I think people will ask about, and I want to nip them in the bud...

I've started #861 to capture ideas of what to mention.

claremacrae commented:
> e.g. the word 'due', 'repeat', 'recurr', 'priority', 'low', 'high' etc, and of course more specific completions -- e.g. various dates and phrases that you can see in the menu when you enter the 'due' or 'recurr' emoji and start typing. A good start would be to use it with length 0 for a while, get used to a few flows (e.g. I'm now very used to type `task description... du<auto-completes> tom<auto-completes>`), and then when you get familiar with what's possible increase the match length to something higher so it would be less obstructive.
> @claremacrae let me know if you think this is something that should be elaborated in the settings description, and if so how.

Sure, I'll have a think about that when writing the docs.
